chained/water_impl.go

```
@@ -42,54 +44,86 @@ func newWaterImpl(dir, addr string, pc *config.ProxyConfig, reportDialCore repor
+			defer wg.Done()
+			log.Debugf("Loading WASM for %q. If not available, it should try to download from the following URLs: %+v. The file should be available here: %s", transport, strings.Split(wasmAvailableAt, ","), dir)
+			vc := newWaterVersionControl(dir)
+			cli := httpClient
+			if cli == nil {
+				cli = proxied.ChainedThenDirectThenFrontedClient(1*time.Minute, "")
+			}
+			downloader, err := newWaterWASMDownloader(strings.Split(wasmAvailableAt, ","), cli)
+			if err != nil {
+				log.Errorf("failed to create wasm downloader: %w", err)
+				return
+			}
+
+			r, err := vc.GetWASM(ctx, transport, downloader)
+			if err != nil {
+				log.Errorf("failed to get wasm: %w", err)
+				return
+			}
+			defer r.Close()
+
+			b, err := io.ReadAll(r)
+			if err != nil {
+				log.Errorf("failed to read wasm: %w", err)
+				return
+			}
+
+			if len(b) == 0 {
+				log.Errorf("received empty wasm")
+				return
+			}
+
+			dialer, err := createDialer(ctx, b, transport)
+			if err != nil {
+				log.Errorf("failed to create dialer: %w", err)
+			}
+			d.dialer = dialer
+		}()
 	}
 
+	return d, nil
+}
+
+func createDialer(ctx context.Context, wasm []byte, transport string) (water.Dialer, error) {
 	cfg := &water.Config{
 		TransportModuleBin: wasm,
 		OverrideLogger:     slog.New(newLogHandler(log, transport)),
 	}
 
 	dialer, err := water.NewDialerWithContext(ctx, cfg)
 	if err != nil {
-		return nil, fmt.Errorf("failed to create dialer: %w", err)
+		log.Errorf("failed to create dialer: %w", err)
 	}
-
-	return &waterImpl{
-		raddr:          addr,
-		dialer:         dialer,
-		reportDialCore: reportDialCore,
-	}, nil
+	return dialer, nil
 }
 
 func (d *waterImpl) dialServer(op *ops.Op, ctx context.Context) (net.Conn, error) {
 	return d.reportDialCore(op, func() (net.Conn, error) {
+		// TODO check if this is the intended behavior or if we should just return an error
+		d.wgDownload.Wait()
+		if d.dialer == nil {
+			return nil, log.Errorf("dialer not available")
+		}
```

Check if what is the intended behavior? There are places where `d.dialer` can be `nil`, so unless I'm missing something, this is correct. Although, we should wrap the error that occurred while trying to download the WASM. That way, we would have a bit more info on why it failed in the traces instead of only in the logs. 
Maybe something like:
```
d.wgDownload.Wait()
if d.dialer == nil {
	d.downloadErrMx.Lock()
	err := d.downloadErr
	d.downloadErrMx.Unlock()

	return nil, log.Errorf("dialer not available: %w", err)
}
```

---

I'll follow the suggestion for adding the error, but I don't believe we need the mutex in this case, since the waitgroup is already saying the operation is done, the error wouldn't be affected by anywhere else.

---

> I don't believe we need the mutex in this case, since the waitgroup is already saying the operation is done
Oh, right. That's a good point.